lib/e2ee/qolmoutboundsession.cpp

```diff
@@ -0,0 +1,128 @@
+// SPDX-FileCopyrightText: 2021 Carl Schwan <carlschwan@kde.org>
+//
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+#include "e2ee/qolmoutboundsession.h"
+#include "e2ee/qolmutils.h"
+
+using namespace Quotient;
+
+QOlmError lastError(OlmOutboundGroupSession *session) {
+    const std::string error_raw = olm_outbound_group_session_last_error(session);
+
+    return fromString(error_raw);
+}
+
+QOlmOutboundGroupSession::QOlmOutboundGroupSession(OlmOutboundGroupSession *session)
+    : m_groupSession(session)
+{
+}
+
+QOlmOutboundGroupSession::~QOlmOutboundGroupSession()
+{
+    olm_clear_outbound_group_session(m_groupSession);
+    delete[](reinterpret_cast<uint8_t *>(m_groupSession));
```

Comment:
wait, i'm not sure i understand your second point - sure, casting it to an array to then delete it looks bad, but why is it a problem if there is other data before it?

Reply:
Hmm, maybe you should scratch that. I just realised that it's actually an obvious pattern to implement _placement_ `new`/`delete` since those don't exist in C. There's no conceivable way for the factory function to return a pointer to a different place than the one we passed to it - the only difference can be in the type. Actually, I'm not sure if there's any advantage over plain calling `malloc` and `free` instead of `new[]` and `delete[]`. But, whatever.

And those people were saying that C++ API doesn't have clear benefits for Olm...